### ISSUE TITLE
Using composition in favor of inheritance in Thing Models?

### ISSUE THREAD

**Issue by thjaeckle** [w3c/wot-thing-description#1177]
I was recently mentioning the well known best practice in computer science "favor composition over inheritance" in an exchange with @sebastiankb and that this is currently hard to achieve with WoT Thing Models.

Currently, Thing Models may "extend" from another Thing Model, or specific elements of a Thing Model may include certain elements of other Thing Models.  
It is however not possible to e.g. define that a Thing Model "Smart Ventilator" consists of:
* a "ventilation" aspect (providing e.g. action "adjustRpm") 
   * extending "switchable" (providing property "on")
* an "illuminance" aspect
   * extending "dimmable" (providing property "dimmer") 
      * extending "switchable" (providing property "on")
* a "climate" aspect (providing properties "humidity", "temperature", "co2level", etc + providing an event "smokeDetected")
* a "firmware" aspect (providing properties describing the running firmware + actions performing a firmware update)

One could - with the currently existing Thing Model - start to model those aspects using the "tm:ref", this however would get quite complicated IMHO.

What I want to start discussing about is whether it would be useful to also allow using composition when modeling Thing Models.  
I am not certain if this would require some other model entity other than "Thing Model" or if this stucture could also be a Thing Model.

Modeling-wise I am thinking in the following direction:
```json
{
    "@context": ["http://www.w3.org/ns/td"], 
    "@type" : "tm:ThingModel", // or something like "tm:ComposingThingModel" ?
    "title": "Smart Ventilator",
    "aspects" : {
        "ventilation" : {
            "tm:ref" :"http://example.com/Ventilator.tm.jsonld" // tm:extends http://example.com/Switchable.tm.jsonld
        },
        "illuminance" : {
            "tm:ref" :"http://example.com/DimmableLamp.tm.jsonld" // tm:extends http://example.com/Dimmable.tm.jsonld tm:extends http://example.com/Switchable.tm.jsonld
        },
        "climate" : {
            "tm:ref" :"http://example.com/EnvironmentSensor.tm.jsonld" // "humidity" (tm:ref including http://example.com/HumiditySensor.tm.jsonld#/properties/humidty), "temperature" (tm:ref including http://example.com/TemperatureSensor.tm.jsonld#/properties/temperature), etc.
        }, 
        "firmware" : {
            "tm:ref" :"http://example.com/FirmwareAware.tm.jsonld"
        }
   }
}
```

With such a structure and having composition combined with inheritance in place, one could for example do such queries against a TD runtime/provider (I hope I use the term correctly):
* give me all runtime thing apsects providing Thing Model "http://example.com/Switchable.tm.jsonld"
   * and set property "on" to `false` to all of them
   * e.g. additionally in combination with a "tag" (e.g. all in room "living room")
* give me all runtime thing apsects providing Thing Model "http://example.com/TemperatureSensor.tm.jsonld"
   * and retrieve their "temperature" value
   * and e.g. perform an "average" aggregation over all selected temperatures
   * this would select both the "Smart Ventilator" as well as a "Weather station" providing the "Temperature" aspect as well

With the inline "tm:ref" inclusion mechanism currently defined, I fear that such queries will get hard to do (because only sub-parts of a model may be included/referenced and not the "complete one").  
If they would be clearly defined in a model, a runtime TD may reference the "composition model" and it would be clear which aspect it provides which will get important when discovering those models during runtime.

Some background where this comes from:
* I am the project lead of the [Eclipse Ditto](https://github.com/eclipse/ditto) project, an OpenSource framework for building digital twins
* Eclipse Ditto lays its main focus on providing APIs and persistence for device abstraction (and therefore could be seen as potential TD implementation in the future)
* a [Thing](https://www.eclipse.org/ditto/basic-thing.html) in Ditto can already be composed of several aspects, so called [Features](https://www.eclipse.org/ditto/basic-feature.html), each of those features may contain one or more `definitions` (links to the model of those runtime aspects)
   * WoT Thing Models are for me a perfect match for a Ditto [Feature](https://www.eclipse.org/ditto/basic-feature.html) (aspect) - however I am missing a modeling part for a Ditto [Thing](https://www.eclipse.org/ditto/basic-thing.html) which defines of which aspects a thing is composed of
* other digital twin standardization initiatives also promote the usage of reusable *submodels* / aspects, see e.g. 
   * [BaSys / BaSyx "Asset Administration Shell"](https://wiki.eclipse.org/BaSyx_/_Documentation_/_AssetAdministrationShell#Submodel_Meta_Model)
   
What do you think?  
Do you think such a submodel / aspect / composition approach in the models is useful or don't you see the need?

**Comment by sebastiankb** [comment]
@thjaeckle many thanks for your input. We haven't considered hierarchy-based things so much yet, and this is a very interesting use case.   

I think, we can simple rely on existing mechanism with the links container and define a specific `rel` type for it (or use an existing one). Here would be an example:

```js
{
    "@context": "https://www.w3.org/2019/wot/td/v1",
    "@type" : "tm:ThingModel", 
    "title": "Smart Ventilator",
    "links" : [
        {
            "rel" : "subthing",
            "href": "http://example.com/Ventilator.tm.jsonld"
        },
        {
            "rel" : "subthing",
            "href": "http://example.com/DimmableLamp.tm.jsonld"
        },
        {
            "rel" : "subthing",
            "href": "http://example.com/EnvironmentSensor.tm.jsonld"
        },
        {
            "rel" : "subthing",
            "href": "http://example.com/FirmwareAware.tm.jsonld"
        }
    ]}
```

I was also thinking about providing an instance specific term in a link entry such as `"instanceTitle" : "ventilation"`.  I think this is not necessary since this can be also clarified within the sub-models like `Ventilator.tm.jsonld` (e.g., by using the root `title`). This is also not an issue when you have two different models that extends the same basic TM like `Switchable.tm.jsonld`. The distinction is then made via the context, which would once be Ventilation and once DimmableLamp.

What do you think?

**Comment by thjaeckle** [comment]
The links could work 👍 
I however don't think that the `"title"` of the referenced models is good enough defining the "role" of the submodel.

For example: imagine modeling a traffic light which is a composite of 3 "Light.tm.jsonld" (red, yellow and green).
I don't think it would be a good idea to model "RedLight.tm.jsonld", "YellowLight.tm.jsonld" and "GreenLight.tm.jsonld" - which I would have to when there is nothing like the `"instanceTitle"` or a `"role"` in the links section.

**Comment by RoboPhred** [comment]
Would subthings be partial things that get merged onto the current thing?  Or would they exist as standalone Things in their own right, enabling Things to exist in a hierarchy?

Assuming they are merging fragments, I'm not sure reusing the same thing model with instancing would be feasible.  Consider bindings; Each affordance specifies its own binding.  If we were to add multiple instances of an identical affordance to a Thing, how would a consumer be able to differentiate the instance it wants in the binding request?  Property gets in particular would have to be able to describe how to pass in the desired instance, possibly by query parameter.  This leads to two problems:
- We would have to supply an instanceId on top of an instanceTitle, as human readable (and possibly translated) titles would make for a lousy identifier.
- The linked aspect/subthing would be required to hold specific annotations in its affordances describing how they can be referenced by instance.  This means the subthing has to be designed from the ground up to be multi-instance compatible.

I also believe it is important that we have the ability to specify a `@type` to be merged into the imported thing's types, so that we can continue to describe the purpose of each instance using json-ld IRIs, just as we can on the affordances.


However, if I can go off topic for a moment, the alternate interpretation of subthings as full standalone non-instanced Things allows us to make Things hierarchical, which is something that is quite useful:
- The Thing Discovery spec is already experimenting with ways of listing groups of things related to a single main thing.  Adding the ability to link things as being children of other things would solve this nicely.
- I have a Scene Manager thing that allows the creation and management of Scene items.  Folding the scenes under the manager would clean up the top-level Thing list, and fit naturally in a thing-hierarchy-aware UI.
- Most of the devices I work with are z-wave.  A z-wave device can have multiple sub-devices called 'endpoints'.  Often, these are multiple instances of identical devices (for example, a plug strip providing an endpoint for each plug).  It would be nice to group these endpoint Things under their primary (endpoint '0') Thing.

Mind you, thing hierarchies seems to be a different issue entirety from what is being asked for here.

**Comment by RoboPhred** [comment]
Thinking on this more, I worry the original post as requested would start to conflict with `@type` annotations we already have.  We have:
```
give me all runtime thing apsects providing Thing Model "http://example.com/Switchable.tm.jsonld"
```
This seems more or less identical to what can already be done with semantic types through `@type`.  iotschema.org in particular already defines a Switch type.  Adding a second method to identify types of things would add more confusion and incompatibilities between systems.

I think the issue is that there are a few interpretations of composition here:
- Composition of semantics, which can already be described by `@type` IRIs.
- Composition of raw definitions, where the parser should take directives to fetch other definitions and merge them into the current definition.
- Composition of clusters of Things, IE thing hierarchies.

The first type should be left as-is.  The second type seems to be the bulk of the request, and needs ironing out with regard to bindings and instancing.  The third would be lovely, but is a different issue altogether.

**Comment by sebastiankb** [comment]
@thjaeckle 
>For example: imagine modeling a traffic light which is a composite of 3 "Light.tm.jsonld" (red, yellow and green).
I don't think it would be a good idea to model "RedLight.tm.jsonld", "YellowLight.tm.jsonld" and "GreenLight.tm.jsonld" - which I would have to when there is nothing like the "instanceTitle" or a "role" in the links section.

I see the point. As human it makes sense to give "Light.tm.jsonld" a kind of context by providing a kind of name or role. However, the "Light.tm.jsonld" itself does not yet know that it now represents "red", "yellow" or "green" and must act in this respect.

I compare this with a programming language:

```js
red = new Light()
yellow = new Light()
green = new Light()
```

As a developer, you must now make sure that the correct color is actually applied in context (e.g., `red.setColor('red')`) before you call `red.on(true)`. This would not the case if we would have 

```js
red = new RedLight()
yellow = new YellowLight()
green = new GreenLight()
```

Each extends the Light class model and now the developer can simple apply `red.on(true)`. 
As alternative we can also introduce constructors like 

```js
red = new Light(RED)
yellow = new Light(YELLOW)
green = new Light(GREEN)
```
That would also mean that we have to provide a kind of placeholder map in the links container. A similar approach that is introduced [here](https://w3c.github.io/wot-thing-description/#thing-model-td-placeholder).

**Comment by sebastiankb** [comment]
@RoboPhred 

>Assuming they are reusable fragments considered part of their linked Thing, I'm not sure reusing the same thing model with instancing would be feasible. Consider bindings; Each affordance specifies its own binding. If we were to add multiple instances of an identical Thing, how would a consumer be able to differentiate between the instance's properties when making a request through a binding? Property get operations in particular would have to be able to describe how to pass in the desired instance. Since this would be a http GET operation, bodies cannot (should not?) be used, leaving us with either a url path or query parameter. 

For me it is only a question how the Thing Description and Sub-Thing Description are generated from hierarchy based Thing Model definitions. E.g., for the traffic light I would expect for the master model traffic light we would have URLs like `http://myTrafficLight.com/properties/status` as example. For the sub-Things I would expect something like `http://myTrafficLight.com/red/properties/on`.

>I also believe it is important that we have the ability to specify a @type against the subthing link, so that we can continue to describe the purpose of each instance using json-ld IRIs, just as we can on the affordances.

@type is always possible to give a semantic / machine readable context. I would also propose to use this in the link container, e.g., 

```js
{
    "@context": "https://www.w3.org/2019/wot/td/v1",
    "@type" : "tm:ThingModel", 
    "title": "Traffic Light",
    "links" : [
        {
            "rel" : "subthing",
            "href": "http://example.com/Light.tm.jsonld",
            "instanceName" : "red",
            "@type" : "schema:redColor"
        },
        {
            "rel" : "subthing",
            "href": "http://example.com/Light.tm.jsonld",
            "instanceName" : "yellow",
            "@type" : "schema:yellowColor"
        },
        {
            "rel" : "subthing",
            "href": "http://example.com/Light.tm.jsonld",
            "instanceName" : "green",
            "@type" : "schema:greenColor"
        }
    ]}
```

>The Thing Discovery spec is already experimenting with ways of listing groups of things related to a single main thing. Adding the ability to link things as being children of other things would solve this nicely.

I have to check this. @farshidtz what do you think?

>I have a Scene Manager thing that allows the creation and management of Scene items. Folding the scenes under the manager would clean up the top-level Thing list, and fit naturally in a thing-hierarchy-aware UI.
Most of the devices I work with are z-wave. A z-wave device can have multiple sub-devices called 'endpoints'. Often, these are multiple instances of identical devices (for example, a plug strip providing an endpoint for each plug). It would be nice to group these endpoint Things under their primary (endpoint '0') Thing.

Interesting, do you have a link to this project? 

>Either way, 'subthing' to me sounds exactly like a hierarchy term, so for linking merging models I think we should go with something else. Possibly "rel": "submodel" or maybe "rel": "blueprint"

I'm kind of open with names. We can also try to reuse existing from IANA like 'item'. We have also included this in the link type list table [here](https://w3c.github.io/wot-thing-description/#link).

**Comment by thjaeckle** [comment]
> To conclude, I would prose to have a kind of 'instanceName'. In addition, there can be @type provided to give a machine interpretable context (e.g., for search/ discovery). The placeholder mentioned may be too advanced for now and maybe something for the TD 2.0.

@sebastiankb This sounds good to me. The example from your last message outlines that quite well.

So to be on the same page here: the `instanceName` can not yet be defined in the `"links"` section, can it? 
But `@type` is already supported as you pointed out in the description of ["link"](https://w3c.github.io/wot-thing-description/#link)?

**Comment by RoboPhred** [comment]
In regards to the thing discovery / hierarchy of thing definitions, there was some discussion about embedding lists of things in other things [here](https://github.com/w3c/wot-discovery/issues/172#issuecomment-844022530).  However, that is definition hierarchy, not composition, so it might not be relevant.


> For me it is only a question how the Thing Description and Sub-Thing Description are generated from hierarchy based Thing Model definitions. E.g., for the traffic light I would expect for the master model traffic light we would have URLs like http://myTrafficLight.com/properties/status as example. For the sub-Things I would expect something like http://myTrafficLight.com/red/properties/on.


Creating a url pattern for such a composition is simple enough, but I worry how it will be described from a binding standpoint.  `http://example.com/Light.tm.jsonld` will have to describe its forms, so how will its forms explain that url pattern to the consumer?    It would need some special identifier for `uriVariables` to indicate where the consumer should place the instance name.

Another thing that worries me about this design is it would require new behavior (code changes) throughout the definition fetching and interaction execution processes if a consumer wants to support this, since it touches pretty much every aspect of a TD interaction lifecycle.  I've been staring at my TD consumption code trying to think how I could fit this in, and I think this might require a ground up redesign since knowing what properties are available and how to use them now require deep knowledge of the thing the property is on, and I would have to track multiple thing descriptions to interact on a single thing.  It might be worth pinging someone from the Scripting API side to see what their take on this is.

The ideal case for composition would be one where the resulting thing description could be 'resolved' back down to a single thing description.  I see a few ways to do this:

## Preprocess the sub-tds into a single composed td

References to other sub-tds get merged directly onto the current TD, so all the work is front-loaded and the resulting thing definition becomes just a standard non-composed TD.

While this still requires TD consumers to treat composition specially, it has the benefit that the work is done ahead of time and the resulting merged TD would be no different from what we have now.
A drawback is that the merger step would have to be complex if we want to support instanceName.  Possibly this could be supported by having placeholders in the sub-td that get resolved during the merger.  For example, the sub-thing provides a form similar to:

http://example.com/Light.tm.jsonld
```json
{
  "properties": {
    "{instanceName}-on": {
      "forms":  [
        {
          "op": "readproperty",
          "href": "{instanceName}/properties/on"
        }
      ]
    }
  }
}
```
During the resolution step, the sub-td would have its properties scanned for `{instanceName}`, to be replaced with the instanceName of the link definition.  Then the property would be merged into the top-level TD for normal consumption.

Essentially, by turning the sub-things into a template or macro model, we can precompile the final TD out of it and treat it no differently than any other non-composing TD.

This process is a little convoluted, and care would need to be taken to support escape characters and such, but it would allow for such composition to be used without drastically changing how things are consumed.

## Use hierarchical things over composition.

This would be my preferred approach for something like the traffic light example.  Rather than trying to compose 3 identical components and finding a way to differentiate them, treat each light as its own thing, and link them from the traffic light.  This requires the least amount of messing with the spec, while at the same time having a more powerful expression.  The only implementation cost would be requiring the consumer to load additional Things from the links (Something it is already capable of doing, considering it managed to load the traffic light).
This is also [already being explored for discovery](https://github.com/w3c/wot-discovery/issues/172#issuecomment-844022530), so there's definitely a need for it.


All in all, I am a bit worried about making the TD spec too complicated to use when it comes to composition.  The cost of implementing a system capable of dealing with composition should be low enough that it sees decent adoption and doesn't get regulated to a few specialty systems.

**Comment by RoboPhred** [comment]
@sebastiankb 

My TD producer/consumer work is over at (WutWot)[https://github.com/RoboPhred/wutwot/tree/develop]
The [Scene Manager](https://github.com/RoboPhred/wutwot/blob/develop/packages/plugins/scenes/src/impl/SceneManagerThingImpl.ts) is a Thing that produces [Scene Things](https://github.com/RoboPhred/wutwot/blob/develop/packages/plugins/scenes/src/impl/SceneThingImpl.ts).

Currently, these are all presented to the UI by a GET request returning an array of items.  I have not yet implemented the Thing Discovery spec for this project, but it is something I plan to do.  Especially if it goes in the direction of a thing hierarchy.

**Comment by mmccool** [comment]
As mentioned in the F2F, it might be interesting to allow "internal" links (e.g. just "#/light") that refer to a standard "definitions" object... that could have members various objects which could include TMs (and other things, such a block could be used for other reusable bits).   Or we could have a special block just for sub-TMs/TDs (e.g. "components", as a collections of TMs/TDs).

**Comment by sebastiankb** [comment]
Composition feature is implemented in TD 1.1. I think we can close this issue.

**Comment by egekorkan** [comment]
Call of 30.11: `tm:submodel` allows this, we can close the issue
